src/vec_common_ppc.h

```diff
@@ -28,14 +28,173 @@
+ * but do not directly support __int128 (39 digit) decimal constants.
+ * Another example is __float128 where the type and Q suffix constants
+ * are recent additions.
+ * In both cases we need to construct; large numeric constants,
+ * special values (infinity and NaN), masks for manipulating the sign
+ * bit and exponent bits.
+ * Often these values will be constructed from vectors of word or
+ * doubleword constants.
+ *
+ * \note GCC does not support expressing an integer constant of type
+ * __int128 for targets where long long integer is less than 128
+ * bits wide.  This applies to the PowerPC target as the long long
+ * type is reserved for 64-bit integers.
+ *
+ * \note GCC __float128 support for the PowerPC target began with
+ * GCC 6. In GCC 6 __float128 support is off by default and has to be
+ * explicitly enabled via the '-mfloat128' option. Starting with GCC 7,
+ * __float128 is enabled by default with VSX support.
+ *
+ * Defining large constants for vectors is
+ * complicated by <I>little-endian</I> (LE) support as
+ * specified in the OpenPOWER ABI and as implemented in the compilers.
+ * Little-endian changes the effective vector element numbering and
+ * the order of constant elements in initializers.  But the __int128
+ * numerical order of magnitude or floating-point format does not
+ * change in registers. The high order bits are on the left and the
+ * low order bits are on the right.
+ *
+ * So for example:
+ * \code
+  const vui32_t signmask = { 0x80000000, 0, 0, 0 };
+  const vui32_t expmask = { 0x7fff0000, 0, 0, 0 };
+ * \endcode
+ * are correct sign and exponent masks for __float128 in big
+ * endian (BE) but would be incorrect for little endian (LE).
+ * To get correct results for both endians, one could code something
+ * like this:
+ * \code
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  const vui32_t signmask = { 0, 0, 0, 0x80000000 };
+  const vui32_t expmask = { 0, 0, 0, 0x7fff0000 };
+#else
+  const vui32_t signmask = { 0x80000000, 0, 0, 0 };
+  const vui32_t expmask = { 0x7fff0000, 0, 0, 0 };
+#endif
+ * \endcode
+ * But this gets tedious after the first dozen times.
+ * Also this can be confusing because it does not appear to the match
+ * the floating-point format diagrams in the PowerISA. The sign-bit
+ * and the exponent are always on the left.
+ *
+ * So this header provides endian sensitive macros that maintain
+ * consistent "magnitude" order. For example:
+ * \code
+  const vui32_t signmask = CONST_VINT128_W (0x80000000, 0, 0, 0);
+  const vui32_t expmask = CONST_VINT128_W (0x7fff0000, 0, 0, 0);
+ * \endcode
+ * This is always correct in either endian.
+ *
+ * Another example; the multiplicative inverse for __int128 10**32
+ * is 211857340822306639531405861550393824741.
+ * The GCC compiler will not accept this constant in a vector __int128
+ * initializer. The next best thing would be
+ * \code
+  // The multiplicative inverse for 1 / 10**32 is
+  // 211857340822306639531405861550393824741
```

Comment:
Should you include the hex version in the comment, so it's clear where the magic hex numbers below come from? 211857340822306639531405861550393824741 = 0x9f623d5a8a732974cfbc31db4b0295e5

Reply:
OK